### PR TITLE
Introduce CachingHiveMetastoreBuilder

### PR DESCRIPTION
**Pull request by sopel39**
* NO RELEASE NOTES *

### CODE REVIEW

**Comment by sopel39** [comment]
This is preparation for caching stats by default

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 1]
```diff
@@ -119,31 +119,89 @@
...
+                .build();
+    }
+
+    public static class CachingHiveMetastoreBuilder
+    {
+        private Optional<HiveMetastore> delegate = Optional.empty();
+        private Optional<Executor> executor = Optional.empty();
+        private OptionalLong expiresAfterWriteMillis = OptionalLong.empty();
+        private OptionalLong refreshMills = OptionalLong.empty();
+        private long maximumSize;
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:142]: Is it safe for this to remain at 0 by default ? Maybe it should be a constructor parameter of the builder

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 2]
```diff
@@ -119,31 +119,89 @@
...
+        return builder()
+                .delegate(delegate)
+                .maximumSize(maximumSize)
+                .statsRecording(StatsRecording.DISABLED)
+                .build();
+    }
+
+    public static class CachingHiveMetastoreBuilder
+    {
+        private Optional<HiveMetastore> delegate = Optional.empty();
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:138]: `delegate` is not an optional field, it must eventually be set.
No point using Optional here.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:142]: moving required stuff to builder construction defeats readability. Builders' advantage is similar to that of "named parameters", so let's not kill that.
it's much cleaner (from API perspective!) to track whether a field was set, eg. using Long (with no default value)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java [hunk 3]
```diff
@@ -208,13 +203,10 @@ public HiveMetastore createMetastore(Optional<ConnectorIdentity> identity)
...
             ConnectorIdentity identity = ConnectorIdentity.ofUser(user);
-            return cachingHiveMetastore(
-                    metastoreFactory.createMetastore(Optional.of(identity)),
-                    new ReentrantBoundedExecutor(executorService, maxMetastoreRefreshThreads),
-                    metastoreCacheTtl,
-                    metastoreRefreshInterval,
-                    metastoreCacheMaximumSize,
-                    metastorePartitionCacheEnabled);
+            return cachingMetastoreBuilder
+                    .delegate(metastoreFactory.createMetastore(Optional.of(identity)))
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java:207]: is `cachingMetastoreBuilder` a shared mutable state?

**Review by findepi — CHANGES_REQUESTED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 1 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:142]: Okay, it should probably be a `OptionalLong` which is unwrapped in the `build` then.

Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java]: I've made `maximumSize` non-optional too

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java (hunk 3 above)
Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java:207]: Yes. I've made builder non-mutable

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 4]
```diff
@@ -119,31 +120,126 @@
...
+        public CachingHiveMetastoreBuilder executor(Executor executor)
+        {
+            CachingHiveMetastoreBuilder builder = duplicate();
+            builder.executor = Optional.of(requireNonNull(executor, "executor is null"));
+            return builder;
+        }
+
+        public CachingHiveMetastoreBuilder cacheTtl(Duration cacheTtl)
+        {
+            CachingHiveMetastoreBuilder builder = duplicate();
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:184]: i like the immutability of this builder, but I am concerned this is now how most of the builders are implemented, and may be confusing to users.
i.e. if i write code that just calls `builder.cacheTtl(..);`, i doubt a reviewer would catch the mistake.

thoughts?

we can at least help a bit annotating the builder methods with `@CheckReturnValue`. A reviewer may not see that, but tooling may

Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:184]: The other option I was thinking was adding `CachingHiveMetastoreBuilder#build(HiveMetastore delegate, Executor executor)` method (these fields changed in `io.trino.plugin.hive.metastore.cache.SharedHiveMetastoreCache#createCachingHiveMetastoreFactory`)?

wdyt?

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:184]: @losipiuk thoughts?

Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:184]: addressed with `CachingHiveMetastoreBuilder builder(CachingHiveMetastoreBuilder other)` method

**Review by findepi — APPROVED** [review]
LGTM

please add `@CanIgnoreReturnValue` to builder configuration methods

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 5]
```diff
@@ -119,31 +121,134 @@
...
+        public CachingHiveMetastoreBuilder partitionCacheEnabled(boolean partitionCacheEnabled)
+        {
+            this.partitionCacheEnabled = partitionCacheEnabled;
+            return this;
+        }
+
+        public CachingHiveMetastore build()
+        {
+            requireNonNull(delegate, "delegate not set");
+            requireNonNull(maximumSize, "maximumSize is set");
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:242]: typo

**Review by raunaqmorarka — APPROVED** [review]

**Pull request merged**